### PR TITLE
Added SDL_AddTimerPrecise() (#11179)

### PR DESCRIPTION
**Pull request by nightmareci**
## Description

Added SDL_AddTimerPrecise()

## Existing Issue(s)

#11179

### CODE REVIEW

File: src/timer/SDL_timer.c [hunk 1]
```diff
@@ -162,17 +183,38 @@ static int SDLCALL SDL_TimerThread(void *_data)
...
             if (interval > 0) {
                 // Reschedule this timer
                 current->interval = interval;
-                current->scheduled = tick + interval;
+                if (current->callback_data.type == SDL_TIMER_CALLBACK_TYPE_PRECISE) {
+		    /* We want good pacing accuracy in the common case, but we
+		       don't want to skip iterations, so only use the
+		       accumulated time if we don't have to skip.
+		     */
+                    if (
```
Comment (slouken) [inline src/timer/SDL_timer.c:207]: spacing?

File: src/timer/SDL_timer.c [hunk 2]
```diff
@@ -201,7 +243,78 @@ static int SDLCALL SDL_TimerThread(void *_data)
...
+
+        precise_timer_done:
+            if (signaled) {
+                current->callback_data.precise.accumulated = 0;
+            } else {
+                current->callback_data.precise.accumulated = (now - start) - delay;
+            }
+            break;
+        }
+        }
```
Comment (slouken) [inline src/timer/SDL_timer.c:317]: We might need a separate thread for precise timers. Right now, an imprecise timer will block a precise one that's scheduled slightly after it.

**Comment by slouken** [comment]
In general, this looks okay. From a user perspective I'm not sure how much "better" a precise timer is, or when I should use it. The caveat that this runs on a separate thread means that precise timing is often combined with other potentially blocking interactions with other threads, and when combined with the normal scheduling caveats means the improvement might not be readily apparent to applications.

Aside from conceptually making it better, do you have a use case in mind for this?

**Comment by nightmareci** [comment]
My original intent was for precise timers to be used for something like game update pacing, where precise timing is preferred, and the snake game's structure seemed an appropriate place to use it. You'd not use precise timers for "non-UI regular tasks", where sloppy timing is fine. After a quick test with the snake game, it seems you're right about how the additional precision doesn't come out in real testing. Though it does seem the fixed timestep logic makes a noticeable difference in average timing. The normal timer in the snake game pretty consistently just overshoots, whereas the precise timer goes above/below to maintain a good average.

Testing with `SDL_AddTimer()`:
Screenshot_20241013_143339_timer

Testing with `SDL_AddTimerPrecise()`:
Screenshot_20241013_143505_timer_precise

So we could just give up on `SDL_AddTimerPrecise()`, but implement the fixed timestep logic for the existing timer types, to get that better average timing. And we might as well implement it for `SDL_AppIterate()`. The extra code in the delay adjustment used for fixed timestepping has negligible performance impact.

**Comment by slouken** [comment]
> So we could just give up on `SDL_AddTimerPrecise()`, but implement the fixed timestep logic for the existing timer types, to get that better average timing. And we might as well implement it for `SDL_AppIterate()`. The extra code in the delay adjustment used for fixed timestepping has negligible performance impact.

That seems reasonable. Do you want to open a new PR for that?

**Pull request closed without merging**